File: Cpp/107_Binary_Tree_Level_Order_Traversal_II.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrderBottom(TreeNode* root) {
        vector<vector<int> > result;
        if(root == nullptr) return result;
        
        queue<TreeNode*> q;
        TreeNode* node = root;
        q.push(node);
        while(!q.empty()){
            int n = q.size();
            vector<int> level;
            for(int i = 0; i < n; ++i){
                node = q.front();
                level.push_back(node->val);
                if(node->left != nullptr) q.push(node->left);
                if(node->right != nullptr) q.push(node->right);
                q.pop();
            }
            result.push_back(level);
        }
        
        std::reverse(result.begin(), result.end());
        
        return result;
    }
};

File: Kotlin/107_Binary_Tree_Level_Order_Traversal_II.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun levelOrderBottom(root: TreeNode?): List<List<Int>> {
        val result: MutableList<MutableList<Int>> = ArrayList()
        if(root == null) return result
        
        val q = LinkedList<TreeNode>()
        q.offer(root)
        
        while(!q.isEmpty()){
            val nodes = mutableListOf<Int>()
            val n = q.size
            for(i in 0 until n){
                val current = q.poll()
                nodes.add(current.`val`)
                if(current.left != null) q.add(current.left)
                if(current.right != null) q.add(current.right)
            }
            result.add(nodes)
        }
        
        return result.reversed()
    }
}